### DISCUSSION TITLE
Automatic Differentiation in Mojo

### DISCUSSION THREAD

**Discussion by jusevitch** [modular/modular#188] [General]
I'd like to know what plans are in place to support automatic differentiation (autodiff) in Mojo. From what I've learned so far, Mojo seems to be well positioned to support some powerful capabilities in this area.

Some of the characteristics I'd love to see in a Mojo-specific autodiff system:
* The flexibility and ease-of-use of both JAX and Julia autodiff libraries
* AOT compilation with support for embedded systems (e.g. putting ML and differentiable programming on small robots / quadrotors / autonomous systems)
* Support for IR-level autodiff similar to [Zygote.jl](https://github.com/FluxML/Zygote.jl)

There are two existing libraries I'm aware of that seem like good matches for Mojo integration. It would be interesting to try these out in the playground at some point:
* [Enzyme](https://enzyme.mit.edu/) - LLVM level autodiff, but it seems like they're working on [MLIR capabilities](https://github.com/EnzymeAD/Enzyme-MLIR-Reverse).
* LAGrad ([paper](https://pengmai.github.io/static/lagrad-paper-preprint.0528666b.pdf), [Github repo](https://github.com/pengmai/lagrad)) - MLIR level autodiff

If Modular has plans to develop an autodiff framework, it would be great to try that out as well.

**Comment by mojodojodev** [comment]
Hi @jusevitch great question, Chris gave an answer to this in the Discord previously:


We don't have any current language features to enable this in the works, but I'm very familiar with the work on tape based and SCT (Source Code Transformation) approaches for AD (automatic differentiation), I believe our metaprogramming features will be useful to explore that when we get to wanting to build it. Tape based approaches will "just work" of course, because they don't need language support.

We also talk to all the existing python apis, including things like nn.module that use this, so those also "just work". I'm not sure if that's the question though

There's a whole body of work called "differentiable programming" that I'm a nerd about 😀, but we don't have immediate plans to work on this. I'm sure we'll intersect with that work in the future.

I can't prove this today, but my intuition is that we will be able to automatically produce backward versions of kernels from the forward version. [This has already been done by a number of other projects](https://enzyme.mit.edu) in other systems and we have strictly more information than those systems do.

**Comment by lattner** [comment]
+1, thank you @mojodojodev !
